File: tiger_ci/tiger_fortran/integral_screen.f90
! This Source Code Form is subject to the terms of the Mozilla Public
! License, v. 2.0. If a copy of the MPL was not distributed with this
! file, You can obtain one at http://mozilla.org/MPL/2.0/.
! all the components need to screen the two-electron integrals using 
! a variety of approaches. 
!
! moved into a module of its own because I need these routines in different 
! cholesky modules and I'm getting tired of dealing with circular dependencies

module integral_screen
  use global_var_mod
  use utilities_mod
  use function_map
  implicit none

  ! Data structure for handling integral screening
  type integral_screen_data
     real(kind=real8), allocatable, dimension(:,:) :: Q_int         ! The Q quantity for CS screening, sqrt[(ij|ij)]
     real(kind=real8), allocatable, dimension(:,:) :: Q_shell      ! The largest sqrt[(ij|ij)] in the Shell pair IJ (i in I, j in J) 
     real(kind=real8) :: tol
     integer :: nSCreenCS
  end type integral_screen_data


contains

  !> \brief This subroutine sets up the data structure used for all integral screening
  !> \param screen The memory for the CS screening quantities
  !> \param nBas  Number of basis functions
  !> \param nShell Number of AO shells
  !> \param diag  The diagonal two electron integrals (ij|ij)
  !> \param bas2shell Maps each AO basis function to its AO shell
  subroutine setup_screening(screen, nBas, nShell, diag, bas2shell)
    implicit none

    ! inputs
    integer, intent(in) :: nBas, nShell
    integer, dimension(:), intent(in) :: bas2shell
    real(kind=real8), dimension(:), intent(in) :: diag
    type(integral_screen_data) :: screen

    ! Hopefully these will get moved to input parameters
    screen%tol   = ao_integral_threshold

    ! Variable initialization
    screen%nSCreenCS  = 0

    call setup_CS_screen(screen, nBas, nShell, diag, bas2shell)
  end subroutine setup_screening




  !> \brief This subroutine sets up the data structure used for Cauchy-Schwarz (CS) screening
  !> \param screen The memory for the CS screening quantities
  !> \param nBas  Number of basis functions
  !> \param nShell Number of AO shells
  !> \param diag  The diagonal two electron integrals (ij|ij)
  !> \param bas2shell Maps each AO basis function to its AO shell
  subroutine setup_CS_screen(screen, nBas, nShell, diag, bas2shell)
    implicit none

    ! inputs
    integer, intent(in) :: nBas, nShell
    integer, dimension(:), intent(in) :: bas2shell
    real(kind=real8), dimension(:), intent(in) :: diag
    type(integral_screen_data) :: screen
    
    ! local variables
    integer :: ierr
    integer :: iBas, jBas
    integer :: iDiag
    integer :: iShell, jShell
    real(kind=real8) :: tmp

    ! allocate memory, if not already allocated 
    if ( .not. allocated(screen%Q_int) ) then
       allocate( screen%Q_int(nBas,nBas), screen%Q_shell(nShell, nShell), stat=ierr)
       call allocatecheck(ierr,"CS integral_screen_data")
    end if
    screen%Q_int = 0.0
    screen%Q_shell = 0.0


    ! First step, building the Q screening quantities at the individual basis set level
    iDiag = 0
    do iBas = 1, nBas
       do jBas = 1, iBas
          iDiag = diag_address(iBas,jBas)
          tmp = sqrt( diag(iDiag) )
          
          screen%Q_int( iBas, jBas ) = tmp
          screen%Q_int( jBas, iBas ) = tmp
       end do
    end do

    ! Now we build the Q screening quantities at the shell level
    do iBas = 1, nBas
       iShell = bas2shell(iBas)
       do jBas = 1, nBas
          jShell = bas2shell(jBas)
          tmp = screen%Q_int(iBas,jBas)
          if ( tmp > screen%Q_shell(iShell,jShell) ) then 
             screen%Q_shell(iShell,jShell) = tmp 
             screen%Q_shell(jShell,iShell) = tmp
          end if
       end do
    end do


  end subroutine setup_CS_screen

  !> \brief Answers the question is this integral shell block screened 
  !> \param data The integral screening data structure
  !> \param iShell  The I in the integral block (IJ|KL) to be screened
  !> \param jShell  The J in the integral block (IJ|KL) to be screened
  !> \param kShell  The K in the integral block (IJ|KL) to be screened
  !> \param lShell  The L in the integral block (IJ|KL) to be screened
  !> \param nFuncInShell  The number of AO functions in each shell
  logical function shellIsScreened(data, iShell, jShell, kShell, lShell, nFuncInShell)
    implicit none
    integer, intent(in) :: iShell, jShell, kShell, lShell
    integer, intent(in) :: nFuncInShell(:)
    type(integral_screen_data) :: data

    shellIsScreened = .false.
    shellIsScreened = isScreenCS(data, iShell, jShell, kShell, lShell, nFuncInShell)

  end function shellIsScreened


  !> \brief Answers the question is this integral shell block screened under CS
  !> \param data The integral screening data structure
  !> \param iShell  The I in the integral block (IJ|KL) to be screened
  !> \param jShell  The J in the integral block (IJ|KL) to be screened
  !> \param kShell  The K in the integral block (IJ|KL) to be screened
  !> \param lShell  The L in the integral block (IJ|KL) to be screened
  !> \param nFuncInShell  The number of AO functions in each shell
  logical function isScreenCS(data, iShell, jShell, kShell, lShell, nFuncInShell)
    implicit none

    ! inputs
    integer, intent(in) :: iShell, jShell, kShell, lShell, nFuncInShell(:)
    type(integral_screen_data) :: data

    ! local variables
    real(kind=real8) :: bound
    integer :: n

    bound = data%Q_shell(iShell,jShell) * data%Q_shell(kShell,lShell)
    isScreenCS = .false.
    if ( bound < data%tol ) then
       iSScreenCS = .true.
       n = nFuncInShell(iShell) * nFuncInShell(jShell) * nFuncInShell(kShell) * nFuncInShell(lShell)
       data%nScreenCS = data%nScreenCS + n
    end if

  end function isScreenCS

  !> \brief Prints out some stats on the number of integrals which weren't calculated due to different screenings
  !> \param data The integral screening data structure
  subroutine print_integral_screen_stats(data)
    implicit none
    type(integral_screen_data) :: data
    write(*,*) "Integrals screened by CS  = ", data%nScreenCS
  end subroutine print_integral_screen_stats

end module integral_screen

File: tiger_ci/tiger_cpp/IntegralScreen.cpp
/*
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/.
 */

/* 
 * @file   IntegralScreen.cpp
 * @author Francis Ricci
 *
 * Created on November 24, 2014, 11:29 AM
 */

#include <iostream>

#include "IntegralScreen.h"
#include "FunctionMap.h"

#include <math.h>

using namespace std;

IntegralScreen::IntegralScreen(double ao_thresh, double cd_thresh, BasisData *data,
                               RestrictedIndex *pair_map){
    _cd_thresh = cd_thresh;
    _ao_thresh = ao_thresh;
    _nScreen = 0;
    _basis_data = data;
    _pair_map = pair_map;
    _nFunc = _basis_data->get_nFuncInShell();
    _idx = new PairIndex(data->get_nBas(), data->get_nBas(), true);
    calc_diag_ints();
    setup_screening();
}

IntegralScreen::~IntegralScreen(){
  delete _idx;
}

/* @brief This subroutine sets up the data structure used for
 *        Cauchy-Schwarz (CS) screening
 * @param diag       The diagonal two electron integrals (ij|ij)
 * @param integrals  An object for getting data bout integrals
 */
void IntegralScreen::setup_screening(){
    //allocate and zero screening arrays
    int nBas = _basis_data->get_nBas();
    int nShell = _basis_data->get_nShell();
    _Q_int.zeros(nBas, nBas);
    _Q_shell.zeros(nShell, nShell);
    
    // First step, building the Q screening quantities at the individual
    // basis set level
    for (int iBas = 0; iBas < nBas; iBas++){
        for (int jBas = 0; jBas <= iBas; jBas++){
            int iDiag = diag_address(iBas, jBas);
            double tmp = sqrt(_diag[iDiag]);
            
            _Q_int(iBas, jBas) = tmp;
            _Q_int(jBas, iBas) = tmp;
        }
    }
    
    vector<int> shell_map(nBas);
    vector<int> nFunc = _basis_data->get_nFuncInShell();
    int base = 0;
    for (int i = 0; i < nShell; i++){
        for (int j = base; j < base + nFunc[i]; j++)
            shell_map[j] = i;
	base += nFunc[i];
    }

    // Now we build the Q screening quantities at the shell level
    for (int iBas = 0; iBas < nBas; iBas++){
        int iShell = shell_map[iBas];
        for (int jBas = 0; jBas < nBas; jBas++){
            int jShell = shell_map[jBas];

            double screen = _Q_int(iBas, jBas);
            if (screen > _Q_shell(iShell, jShell)){
                _Q_shell(iShell, jShell) = screen;
                _Q_shell(jShell, iShell) = screen;
            }
        }
    }
}

/* @brief Answers the question is this integral shell block screened
 * @param i      The I in the integral block (IJ|KL) to be screened
 * @param j      The J in the integral block (IJ|KL) to be screened
 * @param k      The K in the integral block (IJ|KL) to be screened
 * @param l      The L in the integral block (IJ|KL) to be screened
 * @param nFunc  The number of AO functions in each shell
 * @return bool   Is integral shell screened?
 */
bool IntegralScreen::is_screened(int i, int j, int k, int l){
    double bound = _Q_shell(i,j) * _Q_shell(k,l);
    if (bound > _ao_thresh) return false;
    //    _nScreen += _nFunc.at(i) * _nFunc.at(j) * _nFunc.at(k) * _nFunc.at(l);
    return true;
}

/* @brief Has this pair been prescreened out
 * @param i   First member of pair
 * @param j   Second member of pair
 * @return bool   Is this pair screened?
 */
bool IntegralScreen::is_screened(int i, int j){
  int idx = _idx->get_index(i,j);
  if (_prescreened.count(idx) == 0) return false;
  return true;
}
   
    
/* @brief Prints out stats on the number of integrals which weren't
 *        calculated due to CS screening */
void IntegralScreen::print_screen_stats(){
  //    cout << "Integrals screened by CS = " << _nScreen << endl;
}

int IntegralScreen::prescreen(){
    double max_diag = *max_element(_diag.begin(), _diag.end());
    double test = (_cd_thresh * _cd_thresh) / max_diag;
    int nBas = _basis_data->get_nBas();

    /* Now we loop over all the diagonal elements and figure out which (if any)
     * we can safely ignore
     * iDiag points to the diagonal elements 
     * ij    keeps track of how many diagonal elements we have kept so far
     * If an element passes screening we add it to the pairMap
     * otherwise we add it to prescreened
     */
    int ij = 0;
    for (int iBas = 0; iBas < nBas; iBas++){
        for (int jBas = 0; jBas <= iBas; jBas++){
            int iDiag = diag_address(iBas, jBas);
            if (_diag[iDiag] > test)
                _pair_map->set_pair(iBas, jBas, ij++);
            else{
	        int idx = _idx->get_index(iBas, jBas);
                _prescreened.insert(idx);
	    }
        }
    }
    
    cout << " Prescreening threshold      = " << test << endl;
    cout << " Basis pairs                 = " << nBas * (nBas + 1) / 2 << endl;
    cout << " Basis pairs after screening = " << ij << endl;
    cout << endl;
    
    return ij;
}

/* @brief Calculates the diagonal integrals (ij|ij)
 */
void IntegralScreen::calc_diag_ints(){
    vector<int> nFunc = _basis_data->get_nFuncInShell();
    int nShell = _basis_data->get_nShell();
    int nBas = _basis_data->get_nBas();
    _diag.resize(nBas * nBas);
    
    for (int iShell = 0; iShell < nShell; iShell++){
        for (int jShell = 0; jShell <= iShell; jShell++){
            
            //reorder shells for Erkale/libint if necessary
            int erk_iShell = iShell;
            int erk_jShell = jShell;
            
            int am_i = funcs_to_am(nFunc[erk_iShell]);
            int am_j = funcs_to_am(nFunc[erk_jShell]);
            
            if (am_i < am_j){
                int tmp = erk_iShell;
                erk_iShell = erk_jShell;
                erk_jShell = tmp;
            }
            
            vector<double> eris = _basis_data->eval_ijkl(erk_iShell, erk_jShell,
                                                   erk_iShell, erk_jShell);
            
            //We need to place the integrals into the diagonal vector
            //We do this according to Jeremy's original formulas
            for (int iBas = 0; iBas < nFunc[erk_iShell]; iBas++){
                int jLimit = nFunc[erk_jShell] - 1;
                if (erk_iShell == erk_jShell) jLimit = iBas;
                for (int jBas = 0; jBas <= jLimit; jBas++){
                    int i = accumulate(nFunc.begin(), nFunc.begin() + erk_iShell, 0) + iBas;
                    int j = accumulate(nFunc.begin(), nFunc.begin() + erk_jShell, 0) + jBas;
                    
                    int ij = diag_address(i, j);
                    int ijij = ((iBas * nFunc[erk_jShell] + jBas)
                               * nFunc[erk_iShell] + iBas)
                               * nFunc[erk_jShell] + jBas;

                    _diag[ij] = eris[ijij];
                }
            }            
        }
    }
}
